### PR TITLE
Fix NP Mapping

### LINKED ISSUE #371: Bug in re-partitioning with nearest-projection mapping
We somehow lose triangles while re-partitioning. One example is described in #368, another one was reported by @threimann for FEAP. For FEAP, I found that the triangles are not lost during the bounding box filtering, but afterwards during the mapping-based filtering. The normal vectors of some vertices looked suspicious. 
 
We should try to reproduce the problem with an integration test.

### PR DESCRIPTION
**Pull request by fsimonis**
This PR fixes the NP mapping bug and is based on the Hotfix 1.5.1.

This PR will:
* Fix the openness and direction of boost geometry adapters.
* Add tests to the boost geometry adapters, and the generation of interpolation elements.
* Fix the tests of the NP mapping by removing vertex coordinate modification.
* Add additional tests to failing edge cases of the NP mapping.
* Replace the general primitive index by one index per primitive.
  The new version does not require to wrap primitives in bounding boxes, which fixes the encountered bugs.
* The NP mapping now uses the _nearest_ projection instead of using extrapolation.

Fixes #371
Closes #373

### CODE REVIEW

File: src/mapping/NearestProjectionMapping.cpp [hunk 1]
```diff
@@ -133,33 +230,41 @@ void NearestProjectionMapping:: map
...
 
   if (getConstraint() == CONSISTENT){
     for(mesh::Vertex& v : input()->vertices()){
       for (size_t i=0; i < output()->vertices().size(); i++) {
         const InterpolationElements& elems = _weights[i];
         for (const query::InterpolationElement& elem : elems) {
-          if (elem.element->getID()==v.getID() && elem.weight!=0.0) {
+          if (elem.element->getID()==v.getID() && !math::equals(elem.weight,0.0)) {
             v.tag();
+            tagged.insert(v.getID());
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/mapping/NearestProjectionMapping.cpp]: I now use the set for a two-phase tagging.

**Comment by fsimonis** [comment]
After the comment from @floli, I refactored the tagging algorithm.
The old method was `O(nm)`, the new one is `O(n+m)` due to a two-phase scheme:

1) Gather vertices to tag from the weights
2) Tag the vertices

**Comment by uekerman** [comment]
I also looked at the results of the integration test `NearestProjectionRePartitioning` in Paraview. Everything looks good!

File: src/mapping/NearestProjectionMapping.cpp [hunk 2]
```diff
@@ -26,42 +30,135 @@ NearestProjectionMapping:: NearestProjectionMapping
...
-        _weights[i].push_back(elem);
-      }
+
+    const auto &fVertices = from->vertices();
+    const auto &tVertices = to->vertices();
+    const auto &tEdges    = to->edges();
+
+    _weights.resize(fVertices.size());
+
+    constexpr int nnearest = 4;
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/mapping/NearestProjectionMapping.cpp]: Added documentation

File: src/mapping/tests/NearestProjectionMappingTest.cpp [hunk 3]
```diff
@@ -94,64 +124,81 @@ BOOST_AUTO_TEST_CASE(ConsistentNonIncremental2D)
...
+      mapping.map ( inDataID, outDataID );
+      BOOST_TEST ( outData->values()[0] == valueVertex1);
+      BOOST_TEST ( outData->values()[1] == valueVertex2);
+      BOOST_TEST ( outData->values()[2] == (valueVertex1 + valueVertex2) * 0.5 );
+  }
 }
 
 
+/// @TODO FIX broken test
+#if 0
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/mapping/tests/NearestProjectionMappingTest.cpp]: Fixed it
